File: src/main/java/com/hazza/LeetCodeByJava/SegmentTwo/No52NQueensII.java
package com.hazza.LeetCodeByJava.SegmentTwo;

import java.util.HashSet;
import java.util.Set;

/**
 * https://leetcode.com/problems/n-queens-ii/description/
 */
public class No52NQueensII {
    // solution 1
    public int totalNQueens(int n) {
        return dfs(n, 0 , new HashSet<Integer>(), new HashSet<Integer>(), new HashSet<Integer>());
    }

    private int dfs(int n, int row,
                     Set<Integer> xySum, Set<Integer> xyDif, Set<Integer> queens) {
        int count = 0;
        if (row == n) return 1;
        for (int col = 0; col < n; ++col) {
            if (!queens.contains(col) && !xyDif.contains(row - col) && !xySum.contains(row + col)) {
                xySum.add(row + col);
                xyDif.add(row - col);
                queens.add(col);
                count += dfs(n, row + 1, xySum, xyDif, queens);
                xySum.remove(row + col);
                xyDif.remove(row - col);
                queens.remove(col);
            }
        }

        return count;
    }

    // solution 2
    public int totalNQueens1(int n) {
        return dfs1(n, 0, new boolean[n], new boolean[n * 2], new boolean[n * 2]);
    }

    private int dfs1(int n, int row, boolean[] queens, boolean[] xySum, boolean xyDif[]) {
        int count = 0;
        if (row == n) return 1;
        for (int col = 0; col < n; ++col) {
            if (!queens[col] && !xySum[row +col] && !xyDif[n + row - col]) {
                queens[col] = true;
                xySum[row +col] = true;
                xyDif[n + row - col] = true;
                count += dfs1(n, row + 1, queens, xySum, xyDif);
                queens[col] = false;
                xySum[row +col] = false;
                xyDif[n + row - col] = false;
            }
        }
        return count;
    }


}

File: src/main/python/com/hazza/LeetCodeByPython/SegmentTwo/No52_N_Queens_II.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/5/23 20:40
# https://leetcode.com/problems/n-queens-ii/


class Solution:
    def totalNQueens(self, n: int) -> int:
        return self.dfs(n, [], [], [])

    def dfs(self, n, queens, xy_dif, xy_sum):
        cnt, y = 0, len(queens)
        if n == y:
            return 1

        for x in range(n):
            if x not in queens and (
                    x - y) not in xy_dif and (
                    x + y) not in xy_sum:
                cnt += self.dfs(n,
                                queens + [x],
                                xy_dif + [x - y],
                                xy_sum + [x + y])

        return cnt


if __name__ == '__main__':
    solution = Solution()

    assert solution.totalNQueens(4) == 2